walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,153 @@
 
 public final class FindMeetingQuery {
+
+  // The following function returns a Collection<TimeRange> of all possible times for a meeting to occur given 
+  // a Collection<Event> of all known events for that day, and a MeetingRequest. Each event has a title, list of guests attending
+  // this event, and a TimeRange for the event(which includes the start potentialTime and duration of the event). Each MeetingRequest 
+  // has a title, duration requested for the meeting, and a list of guests invited to this meeting. This funtion performs the task by 
+  // starting with the whole day, and then continuosly removing blocks of potentialTime that clash with any other known event, if there is a
+  // guest for the MeetingRequest who has to attend this event.
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> potentialMeetingTimes = new ArrayList<>();
+    // Add the whole day to potential meeting times
+    potentialMeetingTimes.add(TimeRange.WHOLE_DAY);
+
+    Collection<String> mandatoryMeetingGuests = request.getAttendees();
+    Collection<String> optionalMeetingGuests = request.getOptionalAttendees();
+    long meetingDuration = request.getDuration();
+    Collection<TimeRange> finalPossibleTimes = new ArrayList<>();
+
+    ArrayList<TimeRange> unavailableTimesForOptionalGuests = new ArrayList<>();
+    
+    // Check if there are any meeting guests in the event guest list, and if there are, remove the block of time 
+    // for the event from the potential meeting times 
+    for (Event event: events) {
+      TimeRange eventTime = event.getWhen();
+      Set<String> eventGuests = event.getAttendees();
+      // if there are any optional guests attending this event, add the time to the list of unavailable times for optional guests 
+      for (String optionalMeetingGuest: optionalMeetingGuests) {
+        if (eventGuests.contains(optionalMeetingGuest)) {
+          unavailableTimesForOptionalGuests.add(eventTime);
+        }
+      }
+      for (String mandatoryMeetingGuest: mandatoryMeetingGuests) {
+        if (eventGuests.contains(mandatoryMeetingGuest)) {
+          fixSchedulingConflicts(potentialMeetingTimes, eventTime);
+        }
+      }
+    }
+    
+    // Make a copy of good meeting times we have at this point for mandatory guests as a backup, in case scheduling is impossible with optional guests
+    ArrayList<TimeRange> copyOfGoodMeetingTimes = new ArrayList<TimeRange>(potentialMeetingTimes);
+    // Make a hashmap of all eliminated times and frequency of elimination, to optimize scheduling with the most optional guests 
+    HashMap<TimeRange, Integer> eliminatedTimes = new HashMap<>();
+ 
+    for (TimeRange unavailableTime: unavailableTimesForOptionalGuests) {
+      // boolean to check if an unavailable time has already been deleted
+      boolean timeAccounted = false;
```

Comment:
Comments explaining what a variable is for are often a good indication that we could name something more clearly. For example, does `unavilableTimeAlreadyDeleted` let us delete the comment?